### ISSUE TITLE
Build problem

### ISSUE THREAD

**Issue by supnaobed** [afbarnard/go-lbfgsb#2]
When I try to build the project I get the following problem:

`gcc -g -Wall -Wextra -fimplicit-none -finit-local-zero -fbounds-check -std=legacy -Wno-compare-reals -Wno-unused-parameter -c lbfgsb/blas.f -o lbfgsb/blas.o
clang: warning: argument unused during compilation: '-fimplicit-none'
clang: warning: argument unused during compilation: '-finit-local-zero'
clang: warning: argument unused during compilation: '-fbounds-check'`

*warning:* `unknown warning option '-Wno-compare-reals' [-Wunknown-warning-option]`
**error:** `invalid value 'f95' in '-x f95'`
**error:** `invalid value 'legacy' in '-std=legacy'`

**Comment by afbarnard** [comment]
It looks like you are using clang as your compiler instead of gcc (gfortran). (I didn't know clang even compiles Fortran!) Since clang and gcc have different options, you will either have to adjust the makefile so that it works with clang or use gcc/gfortran instead. I don't have any experience with clang, so I can't help you there. Another approach would be to compile the Fortran by hand (without the makefile), and then invoke the makefile to do the rest.

I don't know where your '-x f95' is coming from because it doesn't show up in any of my compile commands, so I suspect that is added by clang.

Have you checked that clang supports Fortran? Your errors and warnings regard Fortran-specific options, which suggests that clang doesn't actually support Fortran.

I would recommend just using the gfortran compiler after checking it is a sufficient version (>= 4.4.6). Specifying gfortran directly will avoid gcc pointing to clang.

    $ gfortran --version
    $ make compiler=gfortran

**Comment by supnaobed** [comment]
Explicitly using gfortran helped start the compilation, but now there is a problem with --relocatable option
```
gfortran -g -Wall -Wextra -fimplicit-none -finit-local-zero -fbounds-check -std=legacy -Wno-compare-reals -Wno-unused-parameter -c lbfgsb/blas.f -o lbfgsb/blas.o
gfortran -g -Wall -Wextra -fimplicit-none -finit-local-zero -fbounds-check -std=legacy -Wno-compare-reals -Wno-unused-parameter -c lbfgsb/linpack.f -o lbfgsb/linpack.o
gfortran -g -Wall -Wextra -fimplicit-none -finit-local-zero -fbounds-check -std=legacy -Wno-compare-reals -Wno-unused-parameter -c lbfgsb/timer.f -o lbfgsb/timer.o
gfortran -g -Wall -Wextra -fimplicit-none -finit-local-zero -fbounds-check -std=legacy -Wno-compare-reals -Wno-unused-parameter -c lbfgsb/lbfgsb.f -o lbfgsb/lbfgsb.o
gfortran -g -Wall -Wextra -fimplicit-none -finit-local-zero -fbounds-check -std=f2003 -c lbfgsb.f03 -o lbfgsb.o
gfortran -g -Wall -Wextra -fimplicit-none -finit-local-zero -fbounds-check -std=f2003 -c lbfgsb_c.f03 -o lbfgsb_c.o
ar cr liblbfgsb.a lbfgsb_c.o lbfgsb.o lbfgsb/lbfgsb.o lbfgsb/blas.o lbfgsb/linpack.o lbfgsb/timer.o
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: liblbfgsb.a(lbfgsb.o) has no symbols
ld --relocatable lbfgsb_c.o lbfgsb.o lbfgsb/lbfgsb.o lbfgsb/blas.o lbfgsb/linpack.o lbfgsb/timer.o -o lbfgsb.syso
ld: unknown option: --relocatable
make: *** [lbfgsb.syso] Error 1
```

**Comment by afbarnard** [comment]
What ld are you running (`ld --version`)? It looks like --relocatable has been in GNU ld for a long time so I suspect you are again getting a substitute, i.e. ld is pointing to clang. What build environment are you using anyway?

In any case, and as far as I remember, `--relocatable` is needed in order to make a system object (*.syso) for Go to include when it compiles your program into a binary, so you will need to find a linker that can do that and then edit the makefile to use that linker. There may be other ways to go about things, like gcc can create a syso directly, so experiment and let me know if you find a build process that works better, or is more direct, or relies on fewer tools/steps, etc.

**Comment by afbarnard** [comment]
How are you coming along with this?

**Comment by supnaobed** [comment]
Unfortunately I was not able to set your library, I think it is not related with the problems of the library, but with my limited knowledge in build projects. Thank you

**Comment by afbarnard** [comment]
Sorry you didn't get it working. If you have any further questions, just reopen this issue.

**Comment by jesand** [comment]
I had a similar problem, which I fixed by changing --relocatable to -r. Apparently ld in Mac OS doesn't support the full name of this parameter.

Here are all the steps I took to get this working on Mac OS:
1. Using Homebrew, install gcc (the current version is 5.2.0)
2. Edit the makefile variables as follows:
  * Change `compiler` to gcc-5
  * Change --relocatable to -r in the ld command
3. Tell the linker where to find libgfortran with an environment variable change:
  * From $GOPATH/github.com/afbarnard/go-lbfgsb run `LIBRARY_PATH=/usr/local/lib/gcc/5 go build`

After this, the library built correctly.
